lib/ex_aws/credentials_ini/file.ex

```diff
@@ -1,17 +1,111 @@
+      sso_start_url sso_region sso_account_id sso_role_name
     )
 
     def security_credentials(profile_name) do
       shared_credentials = profile_from_shared_credentials(profile_name)
       config_credentials = profile_from_config(profile_name)
+      merge_credentials(config_credentials, shared_credentials)
+    end
+
+    defp merge_credentials(
+           %{
+             sso_start_url: sso_start_url,
+             sso_account_id: sso_account_id,
+             sso_role_name: sso_role_name,
+           } = config_credentials,
+           _shared_credentials
+         ) do
+           case get_sso_role_credentials(sso_start_url, sso_account_id, sso_role_name) do
+             {:ok, sso_creds} -> {:ok, Map.merge(sso_creds, config_credentials)}
+             {:error, _} = err -> err
+           end
+    end
+
+    defp merge_credentials(config_credentials, shared_credentials) do
       Map.merge(config_credentials, shared_credentials)
     end
 
+    defp get_sso_cache_file(sso_start_url) do
+      hash = :crypto.hash(:sha, sso_start_url) |> Base.encode16() |> String.downcase()
+
+      System.user_home()
+      |> Path.join(".aws/sso/cache/#{hash}.json")
+    end
+
+    defp get_sso_role_credentials(sso_start_url, sso_account_id, sso_role_name) do
+      sso_start_url
+      |> get_sso_cache_file()
+      |> File.read()
+      |> parse_sso_cache_file
+      |> check_sso_expiration
+      |> request_sso_role_credentials(sso_account_id, sso_role_name)
+      |> rename_sso_credential_keys
+    end
+
+    defp parse_sso_cache_file({:ok, contents}) do
+      contents
+      |> Jason.decode()
+    end
+
+    defp parse_sso_cache_file(err), do: err
+
+    defp check_sso_expiration({:ok, %{"expiresAt" => expires_at_str} = config}) do
+      with {:ok, expires_at, _} <- DateTime.from_iso8601(expires_at_str)
+      do
+        case DateTime.compare(expires_at, DateTime.utc_now()) do
+          :gt -> {:ok, config}
+          _ -> {:error, "SSO access token is expired, refresh the token with `aws sso login`"}
+        end
+      else
+        {:error, err} -> {:error, "SSO cache file has invalid expiration format: #{err}"}
+      end
+    end
+
+    defp check_sso_expiration(err), do: err
+
+    #TODO: Implement check that verifies if a user has aws_sso and source profile configured
+    # Saw this concern raised in saml2aws using aws go sdk and think it could be helpful
+
+    # TODO: the :ex_aws.request() abstraction that can choose hackney or what users set requires signing
+    # We can't sign yet, so this is written to hackney, but need to revisit alternatives
+    defp request_sso_role_credentials({:ok, %{"accessToken" => access_token, "region" => region}}, account_id, role_name) do
+      with {:ok, %{status_code: 200, headers: _headers, body: body}} <-
+             request(
+               :get,
+               "https://portal.sso.#{region}.amazonaws.com/federation/credentials?account_id=#{account_id}&role_name=#{role_name}",
```

Comment:
could potentially use a path builder both here and with the aws sso file path above?

Reply:
Setting this aside for now